### DISCUSSION TITLE
How to run Appium tests in CI without device farm(e.g. Browserstack)

### DISCUSSION THREAD

**Discussion by olegKun** [appium/appium#17788] [Q&A] [answered]
Hi, team!
Firstly, thank you for supporting my country in such difficult times!
Secondly, thanks for creating such a cool and versatile tool for mobile testing. I'm so excited to learn a new direction of automated testing!
My question is as follows: 
The company I work for wants me to provide a proof of concept for mobile automated testing. However, browserstack is an expensive tool that can only be approved after successful POC. So I am asked if I can run appium server, android and iOS simulators on some servers(e.g. aws ec2 instances). I suppose that may be possible for android simulator but I doubt about an iOS one because it requires macOs and XCode installed.

Please let me know, if there is any option or the way I think about what I am asked for is correct.
Thx in advance!

**Comment by jlipps** [comment]
There are some virtual mac farms out there (I've used MacStadium in the past, not sure if they still exist), but yes it can be a bit expensive. Most teams that do local Appium development have their staff using mac hardware or have mac minis on hand.

**Comment by Higgens75** [comment, marked as answer]
We also have a local jenkins master (linux) and some mac minis attached to it locally.

**Reply by olegKun** [reply]
Hi, thx for your response!
So you have Jenkins installed locally? I mean all the test infrastructure is local: CI, appium, android/Ios devices or simulators(on mac mini), am I right?
I mean the cheapest option is to do testing locally?

**Reply by Higgens75** [reply]
yes its all locally. 
Mac minis are connected as nodes to the jenkins master. Appium is installed via CI on the corresponding mac mini node which runs the tests. 
The mac minis can also run Android Emulators,
so you can run all mobile test on Android and iOS Simulators. It's also possible to connect real devices to the mac minis.

**Reply by Higgens75** [reply]
i think its the cheapest in meaning investing money every month, but you also have an invest of work and money to build up the local infrastructure. But if it is finished you will have no costs anymore until apple says your mac minis does get no OS updates anymore and you can't run newest xcode versions anymore. But this takes several years normally.

**Reply by Higgens75** [reply]
@olegKun and if you want the PoC really cheap, you can use the mac of the developer (or yours) to attach it as build node to the jenkins.
